src/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.cs

```diff
@@ -197,6 +197,14 @@ public static bool IsReferenceOrContainsReferences<T>()
             throw new InvalidOperationException();
         }
 
+        /// <summary>
+        /// GetArrayRange helper method for the compiler to slice an array using a range.
+        /// </summary>
+        public static T[] GetArrayRange<T>(T[] array, Range range)
+        {
+            return array.AsSpan(range).ToArray();
```

Comment:
For covariant arrays, this implementation can throw ArrayTypeMismatchException and/or return array of a different type that what it started with. Is this what we want?

Reply:
Maybe something like:
```C#
Type elementType = array.GetType().GetElementType();
int length = GetRangeLength(range);
T[] newArray = (T[])Array.CreateInstance(elementType, length);
array.AsSpan(range).CopyTo(newArray);
return newArray;
```

I realize I've already shared my opinion on this, but I'll just note here again how I think it's unfortunate we're exposing this.  Such a perf trap.
```C#
int[] array = …;
Span<int> span = array[1..3];
```
is going to invoke all that above goo.

Reply:
> I realize I've already shared my opinion on this, but I'll just note here again how I think it's unfortunate we're exposing this

+100

I believe that we would get much better feedback about the design if we did not add these API first and wait for feedback on why there are actually needed and what they are going to use for.

Mentioned it in https://github.com/dotnet/designs/pull/48#discussion_r241552413 , but got no response.